src/main/java/net/dv8tion/jda/bot/sharding/ShardManager.java

```diff
@@ -76,6 +77,62 @@ default void removeEventListener(final Object... listeners)
         this.getShardCache().forEach(jda -> jda.removeEventListener(listeners));
     }
 
+    /**
+     * Adds listeners provided by the listener provider to each shard to the event-listeners that will be used to handle events.
+     * The listener provider gets a shard id applied and is expected to return a listener.
+     *
+     * <p>Note: when using the {@link net.dv8tion.jda.core.hooks.InterfacedEventManager InterfacedEventListener} (default),
+     * given listener <b>must</b> be instance of {@link net.dv8tion.jda.core.hooks.EventListener EventListener}!
+     *
+     * @param  eventListenerProvider
+     *         The provider of listener(s) which will react to events.
+     *
+     * @throws java.lang.IllegalArgumentException
+     *         If the provided listener provider or any of the listeners or provides are {@code null}.
+     */
+    default void addEventListeners(final IntFunction<Object> eventListenerProvider)
+    {
+        Checks.notNull(eventListenerProvider, "event listener provider");
+        this.getShardCache().forEach(jda ->
+        {
+            Object listener = eventListenerProvider.apply(jda.getShardInfo().getShardId());
+            Checks.notNull(listener, "listener");
+            jda.addEventListener(listener);
+        });
+    }
+
+    /**
+     * Remove listeners from shards by their id.
+     * The provider takes shard ids, and returns a collection of listeners that shall be removed from the respective
+     * shards.
+     *
+     * @param eventListenersProvider
+     *        gets shard ids applied and is expected to return a collection of listeners that shall be removed from
+     *        the respective shards
+     *
+     * @throws java.lang.IllegalArgumentException
+     *         If the provided event listeners provider is {@code null}.
+     */
+    default void removeEventListeners(final IntFunction<Collection<Object>> eventListenersProvider)
+    {
+        Checks.notNull(eventListenersProvider, "event listeners provider");
+        this.getShardCache().forEach(jda ->
+            jda.removeEventListener(eventListenersProvider.apply(jda.getShardInfo().getShardId()))
```

Comment:
This does not work when an EventListenersProvider doesn't always return the same listener instance for the a shard id (unless a proper equals() method has been implemented)

Reply:
Also the method names are not consistent: There's `removeEventListenersProvider`, `removeEventListenerProvider` and `removeEventListeners`

Reply:
`removeEventListenersProvider` looks like a typo and should be `removeEventListenerProvider`

As for `ShardManager#removeEventListeners`, can you do a suggestion how it should be called?


>This does not work when an EventListenersProvider doesn't always return the same listener instance for the a shard id (unless a proper equals() method has been implemented)

![image](https://user-images.githubusercontent.com/6048348/36048876-fd2fbba8-0de0-11e8-9e96-47c656061eef.png)
Getting mixed signals here.